azure/services/bastionhosts/spec.go

```diff
@@ -40,69 +41,63 @@ type AzureBastionSpec struct {
 	EnableTunneling bool
 }
 
-// AzureBastionSpecInput defines the required inputs to construct an azure bastion spec.
-type AzureBastionSpecInput struct {
-	SubnetName   string
-	PublicIPName string
-	VNetName     string
-}
-
-// ResourceName returns the name of the bastion host.
-func (s *AzureBastionSpec) ResourceName() string {
-	return s.Name
-}
-
-// ResourceGroupName returns the name of the resource group.
-func (s *AzureBastionSpec) ResourceGroupName() string {
-	return s.ResourceGroup
-}
-
-// OwnerResourceName is a no-op for bastion hosts.
-func (s *AzureBastionSpec) OwnerResourceName() string {
-	return ""
+// ResourceRef implements azure.ASOResourceSpecGetter.
+func (s *AzureBastionSpec) ResourceRef() *asonetworkv1.BastionHost {
+	return &asonetworkv1.BastionHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.Name,
+			Namespace: s.Namespace,
+		},
+	}
 }
 
-// Parameters returns the parameters for the bastion host.
-func (s *AzureBastionSpec) Parameters(ctx context.Context, existing interface{}) (parameters interface{}, err error) {
-	if existing != nil {
-		if _, ok := existing.(armnetwork.BastionHost); !ok {
-			return nil, errors.Errorf("%T is not an armnetwork.BastionHost", existing)
-		}
-		// bastion host already exists
-		return nil, nil
+// Parameters implements azure.ASOResourceSpecGetter.
+func (s *AzureBastionSpec) Parameters(ctx context.Context, existingBastionHost *asonetworkv1.BastionHost) (parameters *asonetworkv1.BastionHost, err error) {
+	bastionHost := &asonetworkv1.BastionHost{}
+	bastionHost.Spec = asonetworkv1.BastionHost_Spec{}
+	if existingBastionHost != nil {
+		bastionHost = existingBastionHost
 	}
 
 	bastionHostIPConfigName := fmt.Sprintf("%s-%s", s.Name, "bastionIP")
-
-	return armnetwork.BastionHost{
-		Name:     ptr.To(s.Name),
-		Location: ptr.To(s.Location),
-		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
-			ClusterName: s.ClusterName,
-			Lifecycle:   infrav1.ResourceLifecycleOwned,
-			Name:        ptr.To(s.Name),
-			Role:        ptr.To("Bastion"),
-		})),
-		SKU: &armnetwork.SKU{
-			Name: ptr.To(armnetwork.BastionHostSKUName(s.Sku)),
-		},
-		Properties: &armnetwork.BastionHostPropertiesFormat{
-			EnableTunneling: ptr.To(s.EnableTunneling),
-			DNSName:         ptr.To(fmt.Sprintf("%s-bastion", strings.ToLower(s.Name))),
-			IPConfigurations: []*armnetwork.BastionHostIPConfiguration{
-				{
-					Name: ptr.To(bastionHostIPConfigName),
-					Properties: &armnetwork.BastionHostIPConfigurationPropertiesFormat{
-						Subnet: &armnetwork.SubResource{
-							ID: &s.SubnetID,
-						},
-						PublicIPAddress: &armnetwork.SubResource{
-							ID: &s.PublicIPID,
-						},
-						PrivateIPAllocationMethod: ptr.To(armnetwork.IPAllocationMethodDynamic),
-					},
+	bastionHost.Spec.AzureName = s.Name
```

Comment:
Do we know if `s.Name` could ever be empty?
If yes, should we construct `bastionHostIPConfigName` only when `s.Name != ""` so that `bastionHostIPConfigName` doesn't end up being `-bastionIP` ?

Reply:
`s.Name` is never empty because of the defaulting webhook here: https://github.com/willie-yao/cluster-api-provider-azure/blob/aso-bastionhosts/api/v1beta1/azurecluster_default.go#L331-L332